### ISSUE TITLE
Add parameter overload for CakeTaskBuilder.Does.

### ISSUE THREAD

**Issue by sqeezy** [cake-build/cake#1370]
I would to use a pattern like the following to dynamically build tasks. 
The problem is that by running ``` .\build.ps1 -target DefaultB ``` I get an null reference for folder in the does delegate. This is possible through closure, isn't it?

```csharp
var folders = new []{"A","B"};
var target = Argument("target","DefaultA");

foreach(var folder in folders)
{
  Task("Default"+folder)
    .Does(folder =>
    {
      CreateDirectory(folder);
      //something else
    });
}

RunTarget(target);
```

**Comment by devlead** [comment]
Typing on phone now but try something like
```cake
var folders = new []{"A","B"};
var target = Argument("target","DefaultA");

Array.ForEach(folders,
    folder=> {
Task("Default"+folder)
    .Does(ctx =>
    {
      CreateDirectory(folder);
      //something else
    });
});

RunTarget(target);
```

**Comment by sqeezy** [comment]
Wow thanks for the fast response. This works.
Still it would be a nicer syntax if we didn't have to use such workarrounds.
Would it be worth the hassle to add this?

**Comment by devlead** [comment]
Well issue is that you're passing folder as an lambda parameter but it's already declared as a string. Which is invalid c#. So wouldn't call it a workaround per say.

Also I choose Array.ForEach because it has better variable scoping, which leads to more deterministic handling of the captured variables in your Does delegate.

But on the go so I might be missing what the task issue is 😄

**Comment by sqeezy** [comment]
My orginal code snippet wasn't quiet right.

```csharp
var folders = new []{"A","B"};
var target = Argument("target","DefaultA");

foreach(var folder in folders)
{
  Task("Default"+folder)
    .Does(()=>
    {
      CreateDirectory(folder);
      //something else
    });
};
```
This gets me a null ref exception for folder. I thought that the compiler makes this work by creating a closure.

**Comment by RichiCoder1** [comment]
I believe the correct way is this:
var folders = new []{"A","B"};
var target = Argument("target","DefaultA");

```foreach(var folder in folders)
{
  var localFolder = folder;
  Task("Default"+folder)
    .Does(()=>
    {
      CreateDirectory(localFolder);
      //something else
    });
};
```

**Comment by dguder** [comment]
After struggling a while with proposal of @RichiCoder1. I found out that only the one from @devlead in the first comment (https://github.com/cake-build/cake/issues/1370#issuecomment-262208959) works  (with 0.17.0). BTW: It works when started from VS Code in debug mode. 
I have no glue why the results are different and I wanted to peek into the created assembly but I don't know how to get it out of cake.

Here is my solution, including adding dependency to default task
```cake// Configuration + Platform
var configurations = new List<Tuple<string, string>>()
{
    Tuple.Create("Release", "Win32"),
    Tuple.Create("ReleaseClient", "Win32"),
    Tuple.Create("ReleaseClient", "x64"),
    Tuple.Create("Debug", "Win32"),
};

var defaultTask = Task("Default");
defaultTask.Task.AddDependency("Clean");
defaultTask.Task.AddDependency("Restore-NuGet-Packages");

// Create tasks based on configurations list
configurations.ForEach(conf =>
{
    var taskName = String.Format("Build-{0}-{1}", conf.Item1, conf.Item2);
    Task(taskName)
        .Does(() => 
        {
            BuildProject(conf.Item1, conf.Item2);
        });
    defaultTask.Task.AddDependency(taskName);
});

void BuildProject(string configuration, string platform)
{
    MSBuild("./Example.csproj", new MSBuildSettings()
        .SetConfiguration(configuration)
        .WithProperty("platform", platform)
        .SetMSBuildPlatform(MSBuildPlatform.Automatic)
        .SetVerbosity(Verbosity.Minimal)
        .SetNodeReuse(false)
    );
}
```

Thanks for this great tool.

**Comment by devlead** [comment]
This is addressed by #1767
